### PR TITLE
Fix IdSchema and PathSchema types

### PR DESCRIPTION
**Pull request by solimant**
### Reasons for making this change

The way `IdSchema` and `PathSchema` are defined results in reaching a conflicting type when the type argument passed to those types is of a primitive type, such as `string`.

For example, in the snippet below ([TS playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAYglhANgEwJLKgXigbwLABQUUAJHMgFxQDOwATnAHYDmA3IQL7sGGiRToAygGMAFhAC2AQwA8AFSxQpjEAD5F8JGgwAyXIWIBtANYQQUJlFMgA9gDMocgLoB+KkLGTZck2aeruLkJecGhNFAAFKWBRRXwiUkYpCQgqWgYWQO4Q-iiYkXFpeUVlNQ0ESOjYvXijawtGKzN7R1cqPNECr3lfEH8s4II+aEF6JmYPQqlFSe708YCoAHoliwA6CDXYCu0oPXmWQeGoUYzmDq7pRQvPIoPmRZX1ze2tDr2aMcOeAie5UJEDDAwCgEDodBsdAohD+oSgAHIcKRyGkvmwoBx4RZqFBGDYQVJqNQ4MwkgAjRDQYA2KDHeGncazaTwtYw1aOOGI5GUT5nVgYrFwHF4glEknkym0ml0+4swjCGyMWgWZCXKRUBksJnTbA4Mg8+HUGwpAD65HhQR+sMggLgwNB4Mh0N+q3+-C5JCSKVRfIF2Nx+KUYtJUgpVOlnM15yqarlLuIbugHq9qV5435mP9IqDxJDYaltM5stZBAVSpBYBjt3VJzRNymcU9yVThuNEBNlZiFu4QA)), when the type argument for `IdSchema` is `string` (i.e. `IdSchema<string>`), the effective resolved type would be an intersection of `FieldId` and `string`, which is impossible:

```ts
type FieldId = {
  $id: string;
};

type IdSchema<T = any> = FieldId & {
  [key in keyof T]?: IdSchema<T[key]>;
};

type FieldPath = {
  $name: string;
};

type PathSchema<T = any> = FieldPath & {
  [key in keyof T]?: PathSchema<T[key]>;
};

type StringIdSchema = IdSchema<string>; // i.e. FieldId & string

type StringPathSchema = PathSchema<string>; // i.e. FieldPath & string

// TypeScript error:
// Type '{ $id: string; }' is not assignable to type 'StringIdSchema'.
//  Type '{ $id: string; }' is not assignable to type 'string'.
const idSchema: StringIdSchema = {$id: 'some_id'};

// TypeScript error:
// Type '{ $name: string; }' is not assignable to type 'StringPathSchema'.
//  Type '{ $name: string; }' is not assignable to type 'string'.
const pathSchema: StringPathSchema = {$name: 'some_path'};
```      

When the type argument is `string`, `keyof string` in TypeScript actually results in `keyof String`, which includes index signatures like `length`, `toString`, etc. However, because `IdSchema` and `PathSchema` map over the keys of the string primitive type, we need to consider that `string` itself is not treated as an object with properties that can be indexed. Thus, `IdSchema` and `PathSchema` will effectively ignore any keys and result in `string`.

This PR addresses this contradiction by limiting the resolved type of `IdSchema` and `PathSchema` to just `FieldId` and `FieldPath`, respectively, when the type argument is a primitive type, in which it leverages the existing type `GenericObjectType`. This would let TypeScript accept the types as intended ([TS playground link](https://www.typescriptlang.org/play/?#code/C4TwDgpgBA4hB2EBOBLAxgeQEYCsJuABVxoBeKAbwFgAoKeqAbXgEMBbCALigGdhV4AcwC63FvBABuWgF9pNWqEhQAYiggAbACYBJLVHLU6UACQot3PgMHy5tRSSh6AymgAWENiwA8hA1HEQAD5-PwgAD2AELR5YBGR0bDwCYkhaegB+VXVtPSgAMkp0hnpGAGsIECgUeCgKkAB7ADMoQmEM7hd3Tx9CcsrhIPkSmWLuNU1dLXkHZQntAAUWYDd-I3oTVg5LfhqbWRmaJWglldcPL19-QJDyMMjo2LhEVExcfCISYqz5rVPVwrrEpMerVWr1Zqtdrcf7nHq+fogQbDBijYzjHJ-ZZuQ7HKDOXZCLoXFj+Ynwqx7IZQAD0NOqADoIAzspM9PYjo4CdZYd0vP5eSTvJShNS6YzmazFtiOXiVA0GqkyJQmgqdtY7ApOXMFeT+eQ9T55YqSGL6SgmSzfnlAVBVQ0OlKplAAD5QACu8C0ECaNQg+k1tHFADkGq0SK5UGBgFBkEgGkhaGgGvA+NUtHCvNxuXtDWszBYoAByHgNDgAfXMRcDNBDYaVkZQ0djSHjiZoydTMbA2MzLGzhMEgp6+a2XGLpYrPZW1cOdfDkEbzbjCaTKbT9sN411Gb5pMMBe4JbLEErWiLABo7WrKIeJyfy-az9XNUA)):

```ts
type GenericObjectType = {
    [name: string]: any;
};

type FieldId = {
  $id: string;
};

type IdSchema<T = any> = T extends GenericObjectType
  ? FieldId & {
      [key in keyof T]?: IdSchema<T[key]>;
    }
  : FieldId;

type FieldPath = {
  $name: string;
};

type PathSchema<T = any> = T extends GenericObjectType
  ? FieldPath & {
      [key in keyof T]?: PathSchema<T[key]>;
    }
  : FieldPath;

type StringIdSchema = IdSchema<string>; // i.e. FieldId

type StringPathSchema = PathSchema<string>; // i.e. FieldPath

type FooType = {foo: string};

type FooIdSchema = IdSchema<FooType>; // i.e. FieldId & { foo?: FieldId | undefined };

// No TypeScript error
const idSchema: StringIdSchema = {$id: 'some_id'};

// No TypeScript error
const pathSchema: StringPathSchema = {$name: 'some_path'};

// No TypeScript error
const fooIdSchema: FooIdSchema = {$id: 'some_id', foo: {$id: 'some_foo_id'}};
```

If this is related to existing tickets, include links to them as well. Use the syntax `fixes #[issue number]` (ex: `fixes #123`).

If your PR is non-trivial and you'd like to schedule a synchronous review, please add it to the weekly meeting agenda: https://docs.google.com/document/d/12PjTvv21k6LIky6bNQVnsplMLLnmEuypTLQF8a-8Wss/edit

### Checklist

- [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://rjsf-team.github.io/react-jsonschema-form/docs/contributing) of the Markdown text I've added
- [X] **I'm adding or updating code**
  - [ ] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [ ] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
- [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by heath-freenome** [comment]
@solimant It looks like your type updates broke the building of `utils`. Can you look into this?

**Comment by heath-freenome** [comment]
@solimant Have you figured out how to fix the build issues?

**Comment by solimant** [comment]
Hi @heath-freenome - I was actually looking into it today. It seems to have to do with how `toIdSchemaInternal` is using keys from the given `schema` param. I plan to share an update today or tomorrow.

**Comment by solimant** [comment]
@heath-freenome - apologies for the delay. Got this fixed on my local. Will wait to see if tests pass.

**Comment by solimant** [comment]
@heath-freenome - could get the tests running when you get a chance? Thank you 🙏🏼

**Comment by solimant** [comment]
@heath-freenome - looks like CI failed again. I had run `npm run build`, and it passed locally. Are there other CI builds I should run locally?

**Comment by solimant** [comment]
@heath-freenome - looks like tests passed this time. The PR branch is out of date, though, which I'm happy to rebase. Let me know how you'd like me to proceed.

**Comment by solimant** [comment]
@heath-freenome - are you good merging this? What are next steps?

File: packages/validator-ajv6/test/utilsTests/getTestValidator.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ export default function getTestValidator<T = any>(options: CustomValidatorOption
     validateFormData(
       formData: T,
       schema: RJSFSchema,
-      customValidate?: CustomValidator<T>,
+      customValidate?: CustomValidator,
```
Comment (heath-freenome) [inline packages/validator-ajv6/test/utilsTests/getTestValidator.ts:25]: @solimant why remove the generic? Did this cause problems?

File: packages/validator-ajv8/test/utilsTests/getTestValidator.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export default function getTestValidator<T = any>(options: CustomValidatorOption
     validateFormData(
       formData: T | undefined,
       schema: RJSFSchema,
-      customValidate?: CustomValidator<T>,
+      customValidate?: CustomValidator,
```
Comment (heath-freenome) [inline packages/validator-ajv8/test/utilsTests/getTestValidator.ts:25]: @solimant why remove the generic? Did this cause problems?

File: packages/utils/src/schema/toPathSchema.ts [hunk 3]
```diff
@@ -78,52 +78,49 @@ function toPathSchemaInternal<T = any, S extends StrictRJSFSchema = RJSFSchema,
...
         if (schemaItems[i]) {
-          pathSchema[i] = toPathSchemaInternal<T, S, F>(
-            validator,
-            schemaItems[i] as S,
-            `${name}.${i}`,
-            rootSchema,
-            element,
-            _recurseList
-          );
+          (pathSchema as { [key in keyof PathSchema<T>]: PathSchema<T> })[i as keyof PathSchema<T>] =
```
Comment (heath-freenome) [inline packages/utils/src/schema/toPathSchema.ts:81]: Does it work to update the `pathSchema` type above? Right now it is (incorrectly?) defined as `PathSchema`. Maybe it wants to be `PathSchema<T>` . Or maybe create a local variable that is cast once to the proper type? I.e.
```
// on line 74 do this?
const keyablePathSchema = pathSchema as { [key in keyof PathSchema<T>]: PathSchema<T> });
```
This avoids multiple casts later and can be commented once as to why its being done?

And then this line and all others below become:

```
keyablePathSchema[i as keyof PathSchema<T>] = ...
```

File: packages/utils/src/schema/toIdSchema.ts [hunk 4]
```diff
@@ -64,7 +64,11 @@ function toIdSchemaInternal<T = any, S extends StrictRJSFSchema = RJSFSchema, F
     for (const name in schema.properties) {
       const field = get(schema, [PROPERTIES_KEY, name]);
       const fieldId = idSchema[ID_KEY] + idSeparator + name;
-      idSchema[name] = toIdSchemaInternal<T, S, F>(
+      (idSchema as { [key in keyof IdSchema<T>]: IdSchema<T> })[name as keyof IdSchema<T>] = toIdSchemaInternal<
```
Comment (heath-freenome) [inline packages/utils/src/schema/toIdSchema.ts:67]: Add a comment here to explain why the cast

**Comment by heath-freenome** [comment]
> @heath-freenome - are you good merging this? What are next steps?

@solimant sorry, I was waiting to review this with @nickgros and we had some additional feedback

**Comment by solimant** [comment]
Thank you for the review and feedback, @heath-freenome  and @nickgros. I'll work on incorporating that.

**Comment by heath-freenome** [comment]
@solimant Thanks! Can you also update the `CHANGELOG.md` file as part of your update?

**Review by solimant — COMMENTED** [review]
@heath-freenome - I have addressed the feedback, and also added an entry to `CHANGELOG.md`. Happy to address any questions.

Regarding the type assertions, I've simplified those. I realized I was not accounting for the array subschema in `PathSchema`, so took care of that. I tried not to alter the logic so that the changes are strictly type changes to reduce the risk.

**Review by heath-freenome — APPROVED** [review]

**Pull request merged**